src/twisted/conch/test/test_ckeygen.py

```diff
@@ -628,3 +629,59 @@ def test_changePassphraseSubtypeV1(self):
         self.assertTrue(
             privateKeyContent.startswith(b"-----BEGIN OPENSSH PRIVATE KEY-----\n")
         )
+
+    def test_useDefaultForKey(self):
+        """
+        L{options} will default to "~/.ssh/id_rsa" if the user doesn't
+        specify a key.
+        """
+
+        def mock_input(*args):
+            return ""
+
+        import builtins
+
+        self.patch(builtins, "input", mock_input)
+        options = {"filename": ""}
+        _getKeyOrDefault(options)
+        self.assertNotEqual(
+            options["filename"],
+            "",
+        )
+        self.assertTrue(
+            "id_rsa" in options["filename"],
+        )
+
+    def test_displayPublicKeyHandleFileNotFound(self):
+        """
+        Ensure FileNotFoundError is handled, whether the user has supplied
+        a bad path, or has no key at the default path.
+        """
+        options = {"filename": "/foo/bar"}
+        with self.assertRaises(SystemExit) as e:
+            displayPublicKey(options)
+            self.assertTrue(
+                "could not be opened, please specify a file." in e,
+            )
```

Comment:
Is the assertion called ?

My suggestion is not to use the context here, and instead make use of the nice feature provided by trial :)

```suggestion
        exception = self.assertRaises(SystemExit, displayPublicKey, options)
        self.assertIn("could not be opened, please specify a file.", e)
```

Reply:
It's not possible to inspect the SystemExit outside of a context manager.  The first line succeeds, but the second is stopped by a TypeError.  
`[ERROR]
Traceback (most recent call last):
  File "/home/eevel/twisted/src/twisted/conch/test/test_ckeygen.py", line 662, in test_displayPublicKeyHandleFileNotFound
    self.assertIn("could not be opened, please specify a file.", e)
  File "/home/eevel/twisted/src/twisted/trial/_synctest.py", line 505, in assertIn
    if containee not in container:
builtins.TypeError: argument of type 'SystemExit' is not iterable
`
Based on stdlib unittest, this test should only succeed if the first line raises a SystemExit, and the second fulfills the assertTrue assertion.

Reply:
maybe

```
self.assertIn("could not be opened, please specify a file.", e.args[0])
```

Reply:
Just to be clear - the reason it "works" inside the `assertRaises` is that it never runs.  The `displayPublicKey` call raises an exception and execution jumps to the `assertRaises` context manager exit handler.

Reply:
Updated.